File: array/MonotonicStack/cpp/MonotonicStack.cpp
#include "MonotonicStack.hpp"

using namespace std;

vector<long long> NextGreaterElements(const vector<long long>& values) {
    vector<long long> answer(values.size(), -1);
    vector<long long> stack;
    stack.reserve(values.size());

    for (int index = static_cast<int>(values.size()) - 1; index >= 0; --index) {
        while (!stack.empty() && stack.back() <= values[index]) {
            stack.pop_back();
        }
        if (!stack.empty()) {
            answer[index] = stack.back();
        }
        stack.push_back(values[index]);
    }

    return answer;
}

File: array/MonotonicStack/java/MonotonicStack.java
public final class MonotonicStack {
    private MonotonicStack() {}

    public static long[] nextGreaterElements(long[] values) {
        long[] answer = new long[values.length];
        java.util.Arrays.fill(answer, -1L);

        long[] stack = new long[values.length];
        int size = 0;

        for (int index = values.length - 1; index >= 0; index--) {
            while (size > 0 && stack[size - 1] <= values[index]) {
                size--;
            }
            if (size > 0) {
                answer[index] = stack[size - 1];
            }
            stack[size++] = values[index];
        }

        return answer;
    }
}
